Sources/PklSwift/ExternalReaderClient.swift

```diff
@@ -0,0 +1,209 @@
+//             options: ExternalReaderClientOptions(
+//                 resourceReaders: [MyResourceReader()]
+//             ))
+//         try await client.run()
+//     }
+// }
+
+public struct ExternalReaderClientOptions {
+    /// Reader to receive requests.
+    public var requestReader: Reader = FileHandle.standardInput
+
+    /// Writer to publish responses.
+    public var responseWriter: Writer = FileHandle.standardOutput
+
+    /// Readers that allow reading custom resources in Pkl.
+    public var moduleReaders: [ModuleReader] = []
+
+    /// Readers that allow importing custom modules in Pkl.
+    public var resourceReaders: [ResourceReader] = []
+
+    public init(
+        requestReader: Reader = FileHandle.standardInput,
+        responseWriter: Writer = FileHandle.standardOutput,
+        moduleReaders: [ModuleReader] = [],
+        resourceReaders: [ResourceReader] = []
+    ) {
+        self.requestReader = requestReader
+        self.responseWriter = responseWriter
+        self.moduleReaders = moduleReaders
+        self.resourceReaders = resourceReaders
+    }
+}
+
+public class ExternalReaderClient {
+    private let moduleReaders: [ModuleReader]
+    private let resourceReaders: [ResourceReader]
+    private let transport: MessageTransport
+
+    public init(options: ExternalReaderClientOptions) {
+        self.moduleReaders = options.moduleReaders
+        self.resourceReaders = options.resourceReaders
+        self.transport = ExternalReaderMessageTransport(
+            reader: options.requestReader, writer: options.responseWriter
+        )
+    }
+
+    public func run() async throws {
+        for try await message in try self.transport.getMessages() {
+            switch message {
+            case let message as InitializeModuleReaderRequest:
+                try await self.handleInitializeModuleReaderRequest(request: message)
+            case let message as InitializeResourceReaderRequest:
+                try await self.handleInitializeResourceReaderRequest(request: message)
+            case let message as ReadModuleRequest:
+                try await self.handleReadModuleRequest(request: message)
+            case let message as ReadResourceRequest:
+                try await self.handleReadResourceRequest(request: message)
+            case let message as ListModulesRequest:
+                try await self.handleListModulesRequest(request: message)
+            case let message as ListResourcesRequest:
+                try await self.handleListResourcesRequest(request: message)
+            case _ as CloseExternalProcess:
+                self.close()
+            default:
+                throw PklBugError.unknownMessage("Got request for unknown message: \(message)")
+            }
+        }
+    }
+
+    public func close() {
+        self.transport.close()
+    }
+
+    func handleInitializeModuleReaderRequest(request: InitializeModuleReaderRequest) async throws {
```

Comment:
Unnecessary async; there's nothing being awaited here

```suggestion
    func handleInitializeModuleReaderRequest(request: InitializeModuleReaderRequest throws {
```